listenerutil/handler.go

```diff
@@ -0,0 +1,101 @@
+package listenerutil
+
+import (
+	"net/http"
+)
+
+type ResponseWriter struct {
+	wrapped http.ResponseWriter
+	// headers contain a map of response code to header map such that
+	// headers[status][header name] = header value
+	// this map also contains values for hundred-level values in the format 1: "1xx", 2: "2xx", etc
+	// defaults are set to 0
+	headers       map[int]map[string]string
+	headerWritten bool
+}
+
+func (w *ResponseWriter) WriteHeader(statusCode int) {
+	w.headerWritten = true
+	w.setCustomResponseHeaders(statusCode)
+	w.wrapped.WriteHeader(statusCode)
+}
+
+func (w *ResponseWriter) Header() http.Header {
+	return w.wrapped.Header()
+}
+
+func (w *ResponseWriter) Write(data []byte) (int, error) {
+	// The default behavior of http.ResponseWriter.Write is such that if WriteHeader has not
+	// yet been called, it calls it with the below line. We will copy that logic so that our
+	// WriteHeader function is called rather than http.ResponseWriter.WriteHeader
+	if !w.headerWritten {
+		w.WriteHeader(http.StatusOK)
+	}
+	return w.wrapped.Write(data)
+}
+
+func (w *ResponseWriter) setCustomResponseHeaders(statusCode int) {
+	sch := w.headers
+	if sch == nil {
+		return
+	}
+
+	// Check the validity of the status code
+	if statusCode >= 600 || statusCode < 100 {
+		return
+	}
+
+	// Setter function to set headers
+	setter := func(headerMap map[string]string) {
+		for header, value := range headerMap {
+			w.Header().Set(header, value)
+		}
+	}
+
+	// Setting the default headers first
+	if val, ok := sch[0]; ok {
+		setter(val)
+	}
+
+	// Then setting the generic hundred-level headers
+	if val, ok := sch[statusCode/100]; ok {
+		setter(val)
+	}
+
+	// Finally setting the status-specific headers
+	if val, ok := sch[statusCode]; ok {
+		setter(val)
+	}
```

Comment:
What is the behavior if multiple different levels specify conflicting header values? Should we use `w.Header().Add` instead in the setter? What happens for duplicates?

Reply:
the latest call to `Header().Set()` overrides previous headers added the same way

Reply:
Is that the behavior we want? I could imagine saying you want both the 4xx and 401 headers to be merged into one.

Reply:
handled offline, this is the behavior we want, since we need overrides to work

Reply:
I also have this concern. Seems like you might want to say "all 2xx level status codes need header X but 201 also needs header Y".

You _could_ duplicate the headers in the config file between 2xx and 201, but I'm not sure why the current behavior is necessary for overrides to work.